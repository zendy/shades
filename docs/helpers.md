# Helpers

Here's a few helpers we've included to make life a bit easier:

- [`style`, for Pseudo-Selectors, Attributes and Property Pattern Matching](./helpers/style)
- [`mq`, for Media Queries](./helpers/mq)
