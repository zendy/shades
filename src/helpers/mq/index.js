import {
  map,
  reduce,
  toPairs,
  fromPairs,
  toLower,
  contains,
  curry,
  mergeAll
} from 'ramda';

import {
  dasherize,
  isString,
  isNumber,
  when,
  betterSet
} from '../../utilities';

const applyUnit = (unit = '') => (original) => (
  original |> when(isNumber).then(value => value.toString() + unit).otherwise(value => value.toString())
)

const transformInputValues = (mapping) => (givenValue) => mapping?.[givenValue] ?? givenValue;

const transformKeywords = (allowedValues, defaultValue) => (givenValue) => {
  if (allowedValues.includes(givenValue |> toLower)) return givenValue;
  return defaultValue;
}

const transformFalseTo = (replacementValue) => (givenValue) => {
  if (givenValue === false) return replacementValue;
  return givenValue;
}

const mqFeatures = {
  screen: {
    accepts: ['media']
  },
  all: {
    accepts: ['media']
  },
  print: {
    accepts: ['media']
  },
  speech: {
    accepts: ['media']
  },
  width: {
    accepts: ['range', 'value'],
    unit: 'px',
    stringify: applyUnit('px')
  },
  height: {
    accepts: ['range', 'value'],
    unit: 'px',
    stringify: applyUnit('px')
  },
  deviceWidth: {
    accepts: ['range', 'value'],
    unit: 'px',
    stringify: applyUnit('px')
  },
  deviceHeight: {
    accepts: ['range', 'value'],
    unit: 'px',
    stringify: applyUnit('px')
  },
  orientation: {
    accepts: ['enum', 'value'],
    values: ['portrait', 'landscape']
  },
  aspectRatio: {
    accepts: ['range', 'value']
  },
  deviceAspectRatio: {
    accepts: ['range', 'value']
  },
  colour: {
    alias: 'color',
    accepts: ['range', 'boolean']
  },
  colourIndex: {
    alias: 'colorIndex',
    accepts: ['range', 'boolean']
  },
  monochrome: {
    accepts: ['range', 'boolean']
  },
  resolution: {
    accepts: ['range', 'value'],
    unit: 'dpi',
    stringify: applyUnit('dpi')
  },
  scan: {
    accepts: ['enum', 'value'],
    values: ['progressive', 'interlace']
  },
  grid: {
    accepts: ['boolean']
  },
  hover: {
    accepts: ['boolean'],
    stringify: transformInputValues({
      true: 'hover',
      false: 'none'
    })
  },
  anyHover: {
    accepts: ['boolean'],
    stringify: transformInputValues({
      true: 'hover',
      false: 'none'
    })
  },
  pointer: {
    accepts: ['value'],
    stringify: transformFalseTo('none')
  },
  anyPointer: {
    accepts: ['value'],
    stringify: transformFalseTo('none')
  },
  update: {
    accepts: ['value'],
    stringify: transformFalseTo('none')
  },
  colourGamut: {
    alias: 'colorGamut',
    accepts: ['value']
  },
  overflowBlock: {
    accepts: ['value'],
    stringify: transformFalseTo('none')
  },
  overflowInline: {
    accepts: ['value'],
    stringify: transformFalseTo('none')
  },
  displayMode: {
    accepts: ['value']
  },
  invertedColours: {
    alias: 'invertedColors',
    accepts: ['boolean'],
    stringify: transformInputValues({
      true: 'inverted',
      false: 'none'
    })
  }
};

const safeJoin = curry(
  (separator, items) => items.filter(Boolean).join(separator)
);

/**
 * Will join any non-falsy values into a single string separated by a space
 * @param  {array<any>} items
 * Items to join
 * @return {string}
 * The result of joining all non-falsy values
 */
const safeJoinSpaced = (...items) => items.filter(Boolean).join(' ');

const wrapFeature = (original) => '(' + original + ')';
const featurePair = (name, value) => wrapFeature([name |> dasherize, value].join(':'));

const stringifyMq = ({ media, features }) => {
  const mediaTypes = media.length && safeJoinSpaced(
    media.join(', '),
    'and'
  );

  const stringifiedFeatures = features.map(([key, value]) => {
    if (value === true) {
      return wrapFeature(key);
    }
    if (value === false) {
      return; // Urrrgghhhhh javascript
    }

    return featurePair(key, value);
  }) |> safeJoin(' and ');

  return safeJoinSpaced(
    '@media',
    mediaTypes,
    stringifiedFeatures
  );
}

const shadesError = (...msgs) => (
  new Error('Shades<helpers.mq>: ' + msgs.join(' '))
);

const transformers = {
  'media': (name) => (combinator) => ({
    [name]: () => (
      combinator.media(name)
    )
  }),
  'boolean': (name, { alias = name, stringify }) => (combinator) => ({
    [name]: (given = true) => combinator.feature([
      alias,
      stringify?.(given) ?? given
    ])
  }),
  'enum': (name, { alias = name, values, stringify }) => (combinator) => {
    const valueMethods = values |> reduce((result, item) => ({
      ...result,
      [item]: () => combinator.feature([
        alias,
        stringify?.(item) ?? item
      ])
    }), {});

    const featureMethod = (givenValue) => {
      if (values |> contains(givenValue |> toLower)) return combinator.feature([
        alias,
        stringify?.(givenValue) ?? givenValue
      ]);
      else throw shadesError('Oops, the', name, 'query received an invalid value!');
    }

    return {
      ...valueMethods,
      [name]: featureMethod
    };
  },
  'value': (name, { alias = name, stringify }) => (combinator) => ({
    [name]: (givenValue) => (
      combinator.feature([
        alias,
        stringify?.(givenValue) ?? givenValue
      ])
    )
  })
}

const pair = curry(
  (first, last) => ([first, last])
);

const rangeCombinator = (prefix) => (combinator) => (data) => {
  const wrapWithPrefix = (original) => [prefix, original].join('-');
  const treatDataAsWidth = isNumber(data);

  if (treatDataAsWidth) {
    const stringyWidth = mqFeatures.width.stringify(data);

    return pair(
      wrapWithPrefix('width'),
      stringyWidth
    ) |> combinator.feature;
  }

  const parsedRange = data |> toPairs |> reduce((result, [featureName, value]) => {
    const featureOptions = mqFeatures?.[featureName];
    const alias = featureOptions?.alias;

    if (featureOptions?.accepts?.includes('range')) {
      const stringedValue = featureOptions?.stringify?.(value) ?? value;
      const nameOrAlias = alias ?? featureName;

      return [
        ...result,
        pair(
          featureName |> wrapWithPrefix,
          stringedValue
        )
      ];
    }

    return result;
  }, []);

  return combinator.feature(...parsedRange);
}

const combine = (...items) => items.filter(Boolean) |> mergeAll;

const rangeFns = (combinator) => ({
  from: rangeCombinator('min')(combinator),
  to: rangeCombinator('max')(combinator)
});

const featureEscapeHatch = (combinator) => (featureName, value) => combinator([featureName, value]);

const mqFactory = (...mediaTypes) => {
  const queryStore = {
    media: betterSet(mediaTypes),
    features: betterSet()
  };

  const addData = {
    media: (mediaName) => {
      queryStore.media.add(mediaName);
      return outerMethods;
    },
    feature: (...extraData) => {
      queryStore.features.add(...extraData);
      return outerMethods;
    }
  }

  const featureMethods = mqFeatures |> toPairs |> reduce((allMethods, [featureName, options]) => {
    const { accepts, alias } = options;

    const newMethods = accepts.reduce((result, item) => {
      const transformerMethod = transformers?.[item];

      if (!transformerMethod) return result;

      const addedTransformerMethods = transformerMethod(featureName, options)(addData);
      const aliasedMethods = alias && transformerMethod(alias, options)(addData);

      return combine(
        result,
        addedTransformerMethods,
        aliasedMethods
      );
    }, {});

    return {
      ...allMethods,
      ...newMethods
    }
  }, {});

  const outerMethods = {
    ...featureMethods,
    ...rangeFns(addData),
    feature: featureEscapeHatch(addData.feature),
    toString: () => stringifyMq({
      features: queryStore.features.toArray(),
      media: queryStore.media.toArray()
    })
  };

  return outerMethods;
}

export default mqFactory;
