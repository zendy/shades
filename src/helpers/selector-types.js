export const PSEUDO_SELECTORS = {
  ELEMENTS: [
    'before',
    'after',
    'backdrop',
    'cue',
    'firstLetter',
    'firstLine',
    'grammarError',
    'placeholder',
    'selection',
    'spellingError'
  ],
  FUNCTIONS: [
    'any',
    'dir',
    'lang',
    'matches',
    'not',
    'nthChild',
    'nthLastChild',
    'nthLastOfType',
    'nthOfType'
  ],
  CLASSES: [
    'active',
    'anyLink',
    'checked',
    'default',
    'defined',
    'disabled',
    'empty',
    'enabled',
    'first',
    'firstChild',
    'firstOfType',
    'fullscreen',
    'focus',
    'focusWithin',
    'hover',
    'indeterminate',
    'inRange',
    'invalid',
    'lastChild',
    'lastOfType',
    'left',
    'link',
    'onlyChild',
    'onlyOfType',
    'optional',
    'outOfRange',
    'placeholderShown',
    'readOnly',
    'readWrite',
    'required',
    'right',
    'root',
    'scope',
    'target',
    'valid',
    'visited'
  ]
};

export const COMBINATORS = {
  COMBINATOR_AND: 'combinator.and',
  COMBINATOR_OR:  'combinator.or',
  PROPERTY_AND:   'property.and',
  PROPERTY_OR:    'property.or'
};

export const COMBINATOR_INSERTS = {
  [COMBINATORS.COMBINATOR_AND]: '&&',
  [COMBINATORS.PROPERTY_AND]:   '&&',
  [COMBINATORS.COMBINATOR_OR]:  '||',
  [COMBINATORS.PROPERTY_OR]:    '||'
};
