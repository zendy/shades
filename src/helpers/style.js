import {
  compose,
  map,
  concat,
  join,
  curry,
  prop,
  propEq,
  equals,
  mapObjIndexed
} from 'ramda';

import styleStore, { getDescriptor } from './style/selector-store';

import {
  dasherize,
  isString,
  isSymbol,
  isNumber,
  isArray,
  isDefined,
  when,
  betterSet,
  includes,
  stateful,
  proxyFunction,
  proxyPropertyGetter,
  proxyRecord,
  anyOf,
  allOf,
  firstItem,
  lastItem
} from '../utilities';

import {
  PSEUDO_SELECTORS,
  COMBINATORS,
  COMBINATOR_INSERTS
} from './selector-types'

const isDescriptorSym    = Symbol('Compute Selector');
const isDescriptor       = (value) => value?.[isDescriptorSym] ?? false;

const keyFromSymbol      = (...args) => Symbol.keyFor(...args);
const symbolFromKey      = (...args) => Symbol.for(...args);
const styleStoreKey = Symbol.for('Shades: styleStore');

const asPseudoClass      = (name) => `:${name |> dasherize}`;
const asPseudoElement    = (name) => `::${name |> dasherize}`;
const asPropertySelector = (givenName) => `!!${givenName}`;

const asPseudoFunction = curry((name, value) => (
  `:${name |> dasherize}(${value})`
));

const storeDescriptor = (descriptorItem) => {
  const symbolKey = descriptorItem.symbolKey;

  styleStore.lift(
    ({ addItem }) => addItem(symbolKey, descriptorItem)
  );

  return descriptorItem;
};

const createDescriptor = (kind, config = {}) => (value) => {
  const keyOrValue = (
    config?.stringKey ?? value
    |> when(isString).otherwise(JSON.stringify)
  );

  const symbolKey  = symbolFromKey(keyOrValue);

  const selfDescriptor = {
    [isDescriptorSym]:    true,
    kind,
    value,
    symbolKey,
    originalKey:          keyOrValue,
    // this toString doesnt actually return a string
    // it instead returns a symbol whose key is a string
    toString:        () => symbolKey
  };

  return selfDescriptor;
};

const toString = (value) => value.toString();

const descriptorToString = when(isDescriptor).then(prop('originalKey'));
const anythingToString = when(isString).otherwise(toString);

const createAndStoreDescriptor = (kind, config) => compose(
  storeDescriptor,
  createDescriptor(kind, config)
);

const createCombinator = (kind) => (...data) => {
  const stringKey = (
    data
    |> map(
      compose(
        anythingToString,
        descriptorToString
      )
    )
    |> join(` ${COMBINATOR_INSERTS[kind]} `)
  );

  return data |> createAndStoreDescriptor(
    kind,
    { stringKey }
  );
}

const pseudoCombinators = {
  and: createCombinator(
    COMBINATORS.COMBINATOR_AND
  ),
  or:  createCombinator(
    COMBINATORS.COMBINATOR_OR
  )
};

const withAttribute = (givenName) => {
  const quoteString = JSON.stringify;

  const attrWithValue  = (givenValue) => `[${givenName}=${givenValue  |> quoteString}]`;
  const attrStartsWith = (givenValue) => `[${givenName}^=${givenValue |> quoteString}]`;
  const attrEndsWith   = (givenValue) => `[${givenName}$=${givenValue |> quoteString}]`;
  const attrContains   = (givenValue) => `[${givenName}*=${givenValue |> quoteString}]`;

  const anyCombinator = (mapperFn) => (...givenValues) => pseudoCombinators.or(
    ...givenValues.map(mapperFn)
  );

  const plainValue = `[${givenName}]`;

  const outerMethods = {
    anyOf:         anyCombinator(attrWithValue),
    contains:      attrContains,
    containsAny:   anyCombinator(attrContains),
    startsWith:    attrStartsWith,
    startsWithAny: anyCombinator(attrStartsWith),
    endsWith:      attrEndsWith,
    endsWithAny:   anyCombinator(attrEndsWith),
    toString: () => plainValue
  };

  return proxyFunction(
    attrWithValue,
    outerMethods
  )
}

const pseudoClassHandler = (specialChains) => (targetName) => {
  if (PSEUDO_SELECTORS.CLASSES.includes(targetName)) return (
    targetName |> asPseudoClass
  );

  if (PSEUDO_SELECTORS.FUNCTIONS.includes(targetName)) return (
    compose(
      asPseudoFunction(targetName),
      descriptorToString
    )
  );

  if (PSEUDO_SELECTORS.ELEMENTS.includes(targetName)) return (
    targetName |> asPseudoElement
  )

  return specialChains?.[targetName];
}



const style = (
  proxyPropertyGetter(
    pseudoClassHandler({
      ...pseudoCombinators,
      element: proxyPropertyGetter(asPseudoElement, asPseudoElement),
      pseudo: (name, value) => {
        if (isDefined(value)) return asPseudoFunction(name, value);
        return asPseudoClass(name);
      },
      data: proxyPropertyGetter(
        when(isString).then(compose(
          withAttribute,
          concat('data-'),
          dasherize
        ))
      ),
      attr: proxyPropertyGetter(
        when(isString).then(compose(
          withAttribute,
          dasherize
        ))
      ),
      prop: proxyPropertyGetter(
        when(isString).then(asPropertySelector)
      ),
      props: ({
        any: createCombinator(COMBINATORS.PROPERTY_OR),
        all: createCombinator(COMBINATORS.PROPERTY_AND)
      })
    })
  )
)

export default style;
