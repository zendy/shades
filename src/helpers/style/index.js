import {
  createStyleHelpers
} from './style-helper';

export default createStyleHelpers({ useProxyFeatures: true });
