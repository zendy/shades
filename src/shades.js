import { css } from './style-parser';
export default css;
