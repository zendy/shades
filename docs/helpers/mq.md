# `mq` (Media Queries)

The `mq` helper is a nice DSL for writing media queries.  It simply lets you use chained methods to create media query strings, which you can then use in your shades styles.  Note, you can also write media queries without this helper if you wish, as demonstrated in the [Readme example](/) of the `Button` component :)

Example:

```js
import React from 'react';
import shades from '@bupa-digital/shades/react';
import { mq, style } from '@bupa-digital/shades/helpers';

export const ResponsiveButton = shades.button({
  border: '1px solid',
  [style.prop.dark]: {
    borderColor: colours.button.border.dark
  },
  // Here, the result of this chain will be:
  // `@media screen and (min-width:300px) and (max-width:750px)`
  [mq().screen().from(300).to(750)]: {
    border: '1px dotted',
    [style.active]: {
      border: '2px dotted'
    }
  }
})
```

The mq helper supports all valid media features and types from the CSS4 spec (as far as we know).  There are different ways each media feature can be used based on the specification:

- Range (features that can be specified as `min-` or `max-`)
- Boolean (features that are implicit booleans and need not be given a value)
- Enum (features that only allow certain globally-unique values)
- Value (features that can be given any value as-is)

## Range

Range features can be specified using the `from` and `to` methods.  Any feature that can be specified as `min` or `max` is a "range" feature.  The most commonly used range feature is `width`, so if you pass only a number to the range methods (or a string), it will be treated as a width value in pixels, e.g: `mq().from(300).to(700)` or `mq().from('300px').to('700px')` both output `(min-width:300px) and (max-width:700px)`.

You can specify other range values using an object literal.  This also allows you to specify multiple range values at once.  Example:

```js
mq().from({
  height: 700,
  aspectRatio: '2/5'
}).to({
  height: 1200,
  aspectRatio: '7/13'
});
// @media (min-height:700px) and (min-aspect-ratio:2/5) and (max-height:1200px) and (max-aspect-ratio:7/13)
```

> Note: Some features support [Magic Numbers](#magic-numbers), which add a default unit to plain number values.

Range features are:

- `width`
- `height`
- `deviceWidth`
- `deviceHeight`
- `aspectRatio`
- `deviceAspectRatio`
- `colour`
- `colourIndex`
- `monochrome`
- `resolution`

## Enum

There are only 2 enum features (because their values are unique enough for this type), `orientation` and `scan`.  The allowed values for enums have their own methods:

```js
mq().orientation('portrait');
// is the same as
mq().portrait();
// both output: @media (orientation:portrait)
mq().scan('interlace');
// is the same as:
mq().interlace();
// both output: @media (scan:interlace)
```

For orientation, you can call either `portrait` or `landscape` as methods.  For scan, there's `interlace` or `progressive`.

## Boolean

Features that can be implicit booleans can be called as methods without any arguments.  For example:

```js
mq().colour();
// @media (color)
```

> Note: Any feature methods which include the word "colour" will output "color" in order to match the spec.  You can also call the methods using either spelling, but obviously in our examples, we'll prefer the *correct* english version ;)

Boolean features are:

- `colour`
- `colourIndex`
- `monochrome`
- `grid`
- `hover`
- `anyHover`
- `invertedColours`

### Boolean keywords

There are some features which are effectively booleans (by taking either `none` or one specific value).  You can treat these features like other implicit booleans, by just calling the method with no value (which will set the true value), or passing `false` to set it to `none`.  You can even just pass the values yourself if you'd prefer.

Boolean keyword features are:

- `hover`:
  - true: `hover`
  - false: `none`
- `anyHover`:
  - true: `hover`
  - false: `none`
- `invertedColours`:
  - true: `inverted`
  - false: `none`

## Value

Features that can have values can be called as methods, passing the desired value as the only argument.  For example:

```js
mq().width('500em').height('100px');
// @media (width:500em) and (height:100px)
```

### `false` transformation:

Any media feature that can be set to `none` can be given a value of `false`, which will be transformed to `none` in the output.  Features that support this are:

- `pointer`
- `anyPointer`
- `update`
- `overflowBlock`
- `overflowInline`

## Magic Numbers

Magic Numbers are plain number values that, when supplied to features that support them, will be given a default unit.  Default units (and their supported media features) are:

- `px`:
  - `width`
  - `height`
  - `deviceWidth`
  - `deviceHeight`
- `dpi`:
  - `resolution`

## Media Types

There are 4 media types that can be called as methods as well.  They are:
- `screen`
- `print`
- `speech`
- `all`

Like implicit booleans, you just call the method without passing an argument, like so:

```js
mq().print().height(300);
// @media print and (height:300px)
mq().screen().speech().height(300);
// @media screen, speech and (height:300px)
```

You can also pass media types as arguments to the `mq` function itself, if you prefer:

```js
mq('screen', 'print').height(300);
// @media screen, print and (height:300px)
```

## Escape hatch

We've included an escape-hatch for cases where we may be missing a valid media feature, or you want to use an unofficial or experimental feature:

```js
mq().feature('gigantic-manatee', 'underwater');
// @media (gigantic-manatee:underwater)
```

## Operators

Unfortunately we don't currently have support for operators such as `not` and `or`.  I hope to add these in the future.

## More examples...

For more concrete usage examples, please take a look at the `mq.spec.js` file.
