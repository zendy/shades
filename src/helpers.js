import {
  map,
  reduce,
  toPairs,
  fromPairs,
  toLower,
  contains,
  curry,
  mergeAll
} from 'ramda';

import {
  dasherize,
  isString,
  isNumber,
  when,
  msg
} from './utilities';

import mq from './helpers/mq';

export const states = do {
  const normalFn = (selectors) => {
    return Object.entries(selectors).reduce((result, [key, value]) => ({
      ...result,
      [`:${dasherize(key)}`]: value
    }), {});
  }

  normalFn.all = (...selectors) => (styleRules) => selectors |> reduce((result, currentSelector) => ({
    ...result,
    [`:${dasherize(currentSelector)}`]: styleRules
  }), {});

  msg().deprecated('states', normalFn);
}

export {
  mq
}
