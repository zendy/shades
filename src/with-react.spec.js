import React from 'react';
import { mount, render } from 'enzyme';

import shades, { Shades } from './with-react';
import { states, mq } from './helpers';

describe('Shades DOM', () => {
  const mountShades = (data) => mount(
    <Shades to={document.querySelector('head')}>
      {data}
    </Shades>
  );

  it('Renders without incident', () => {
    const Title = shades.h1({
      backgroundColor: {
        dark: 'blue',
        light: 'green',
        default: 'pink'
      },
      fontWeight: {
        dark: 600,
        light: 200
      },
      color: 'black',
      [mq().screen().from(300).to(750)]: {
        backgroundColor: 'purple'
      }
    });

    const titleSubject = mountShades(
      <Title dark>Hello</Title>
    );

    expect(titleSubject).toMatchSnapshot();
  });
  it('renders a block matcher without incident', () => {
    const Linky = shades.a({
      background: 'purple',
      fontWeight: {
        dark: 600,
        light: 200
      }
    }).match({
      dark: {
        color: 'white',
        background: 'green'
      },
      light: {
        color: 'purple',
        background: 'white'
      }
    });

    const darkSubject = mountShades(
      <Linky dark>Hello</Linky>
    );
    expect(darkSubject).toMatchSnapshot();
    expect(darkSubject.find('a').exists()).toBeTruthy();

    const noPropsSubject = mountShades(
      <Linky>No Props!</Linky>
    );
    expect(noPropsSubject).toMatchSnapshot();
    expect(noPropsSubject.find('a').exists()).toBeTruthy();
  });

  it('Forwards valid DOM props', () => {
    const Linky = shades.a({
      backgroundColor: {
        dark: 'blue',
        light: 'green',
        default: 'pink'
      },
      fontWeight: {
        dark: 600,
        light: 200
      },
      color: 'black'
    });

    const subject = mountShades(
      <Linky href="hello.html" superDark="yeah" data-testing="just a test" aria-label="hello">Hello</Linky>
    );

    const anchorItem = subject.find('a');

    expect(anchorItem).toHaveProp('href', 'hello.html');
    expect(anchorItem).toHaveProp('data-testing', 'just a test');
    expect(anchorItem).toHaveProp('aria-label', 'hello');
    expect(anchorItem).not.toHaveProp('superDark');
  });
});
