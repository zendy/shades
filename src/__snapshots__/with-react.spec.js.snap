// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shades DOM Renders without incident 1`] = `
<withContext(Shades.Provider)
  prefixer={true}
  to={
    <span>
      <style
        data-shades=""
      />
    </span>
  }
>
  <Shades.Provider
    prefixer={true}
    to={
      <span>
        <style
          data-shades=""
        />
      </span>
    }
  >
    <shades.h1
      dark={true}
    >
      <h1
        className="shades-h1-14571595172437 shades-element-10496356530578"
      >
        Hello
      </h1>
    </shades.h1>
  </Shades.Provider>
</withContext(Shades.Provider)>
`;

exports[`Shades DOM allows extending multiple generic components 1`] = `
".shades-a-650341851635 {
  font-size: 12px;
  color: blue;
}

.shades-a-650341851635:hover {
  border: 1px solid #ccc;
  text-decoration: underline;
}"
`;

exports[`Shades DOM allows extending multiple generic components 2`] = `
".shades-a-13747910882693 {
  font-size: 12px;
  color: red;
}

.shades-a-13747910882693:hover {
  border: 1px solid #ccc;
  -webkit-text-decoration: dotted;
  text-decoration: dotted;
}"
`;

exports[`Shades DOM properly merges extended style rules 1`] = `
".shades-span-17154989453634 {
  background: black;
  border: 2px solid purple;
  color: white;
  font-weight: 600;
  text-transform: uppercase;
}

.shades-span-17154989453634:hover {
  text-decoration: underline;
  border-color: green;
}"
`;

exports[`Shades DOM renders a block matcher without incident 1`] = `
<withContext(Shades.Provider)
  prefixer={true}
  to={
    <span>
      <style
        data-shades=""
      />
    </span>
  }
>
  <Shades.Provider
    prefixer={true}
    to={
      <span>
        <style
          data-shades=""
        />
      </span>
    }
  >
    <div>
      <shades.a
        dark={true}
      >
        <a
          className="shades-a-17361503962046 shades-element-13128135507417"
        >
          Original Linky
        </a>
      </shades.a>
      <shades.a
        dark={true}
      >
        <a
          className="shades-a-17361503962046 shades-element-13128135507417"
        >
          DUPLICATE linky
        </a>
      </shades.a>
    </div>
  </Shades.Provider>
</withContext(Shades.Provider)>
`;

exports[`Shades DOM renders a block matcher without incident 2`] = `
<withContext(Shades.Provider)
  prefixer={true}
  to={
    <span>
      <style
        data-shades=""
      />
    </span>
  }
>
  <Shades.Provider
    prefixer={true}
    to={
      <span>
        <style
          data-shades=""
        />
      </span>
    }
  >
    <shades.a
      dark={true}
    >
      <a
        className="shades-a-17361503962046 shades-element-13128135507417"
      >
        Hello
      </a>
    </shades.a>
  </Shades.Provider>
</withContext(Shades.Provider)>
`;

exports[`Shades DOM renders a block matcher without incident 3`] = `
<withContext(Shades.Provider)
  prefixer={true}
  to={
    <span>
      <style
        data-shades=""
      />
    </span>
  }
>
  <Shades.Provider
    prefixer={true}
    to={
      <span>
        <style
          data-shades=""
        />
      </span>
    }
  >
    <shades.a>
      <a
        className="shades-a-8244443886614 shades-element-13128135507417"
      >
        No Props!
      </a>
    </shades.a>
  </Shades.Provider>
</withContext(Shades.Provider)>
`;

exports[`Shades DOM should allow styles to apply to custom components 1`] = `
<withContext(Shades.Provider)
  prefixer={true}
  to={
    <span>
      <style
        data-shades=""
      />
    </span>
  }
>
  <Shades.Provider
    prefixer={true}
    to={
      <span>
        <style
          data-shades=""
        />
      </span>
    }
  >
    <shades.generic>
      <Component
        className="shades-generic-13196722402662 shades-element-2116950705577"
      >
        <div
          className="shades-generic-13196722402662 shades-element-2116950705577"
        >
          Hello this is a test
        </div>
      </Component>
    </shades.generic>
  </Shades.Provider>
</withContext(Shades.Provider)>
`;
